### ISSUE TITLE
For AWS Credentials , can we use AWSCredentialsProviderChain . This will provide the credentials if the application running on EC2, ECS, EKS ,etc or from default env variables.

### ISSUE THREAD

**Issue by sankalpkale88** [googleapis/google-auth-library-java#714]
https://github.com/googleapis/google-auth-library-java/blob/aa7ede1d1c688ba437798f4204820c0506d5d969/oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java#L272

https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/AWSCredentialsProviderChain.html

**Comment by sankalpkale88** [comment]
@VisibleForTesting
  AwsSecurityCredentials getAwsSecurityCredentials() throws IOException {

    AWSCredentialsProvider awsCredentialsProvider = DefaultAWSCredentialsProviderChain.getInstance();
    AWSCredentials awsCredentials = awsCredentialsProvider.getCredentials();
    if (awsCredentials instanceof BasicSessionCredentials){
      BasicSessionCredentials basicSessionCredentials = (BasicSessionCredentials) awsCredentials;
      return new AwsSecurityCredentials(basicSessionCredentials.getAWSAccessKeyId(),
              basicSessionCredentials.getAWSSecretKey(),basicSessionCredentials.getSessionToken());
    }
    return new AwsSecurityCredentials(awsCredentials.getAWSAccessKeyId(), awsCredentials.getAWSSecretKey(),null);
  }

**Comment by oprigan-cgi** [comment]
We have the same problem when using the lib in AWS Lambda. 

AWS Lambda is using an assumed IAM Role with temporary security credentials and therefore It is not possible to create a valid subject token to be exchanged with sts.googleapis.com.

In AWSCredentials the following code exists:
```
  @VisibleForTesting
  AwsSecurityCredentials getAwsSecurityCredentials() throws IOException {
    // Check environment variables for credentials first.
    String accessKeyId = getEnvironmentProvider().getEnv("AWS_ACCESS_KEY_ID");
    String secretAccessKey = getEnvironmentProvider().getEnv("AWS_SECRET_ACCESS_KEY");
    String token = getEnvironmentProvider().getEnv("Token");
    if (accessKeyId != null && secretAccessKey != null) {
      return new AwsSecurityCredentials(accessKeyId, secretAccessKey, token);
    }
    ...

```  
As @sankalpkale88 mentioned the provider chain could be a possible solution.

On the other hand the environment variable for the session token in an AWS environment is named "**AWS_SESSION_TOKEN**" and not "Token" so this could be another approach to fix the issue.

```
    String token = getEnvironmentProvider().getEnv("AWS_SESSION_TOKEN");
```

Kind regards,
Oliver

**Comment by TimurSadykov** [comment]
@oprigan-cgi Could you please confirm if this issue now resolved? Given the one of the possible solutions got implemented.

**Comment by pwalczak** [comment]
Hi team!

Is there some plan to implement this? This would allow services deployed on AWS EKS using IRSA to get successfully the GCP tokens using WIF.

Thanks!
Piotrek

**Comment by TimurSadykov** [comment]
@pwalczak please clarify what exactly are you referring to? The AWSCredentialsProviderChain or something else? The original issue got mitigated by a related fix: https://github.com/googleapis/google-auth-library-java/issues/723

**Comment by pwalczak** [comment]
Hi @TimurSadykov. Yes, the AWSCredentialsProviderChain. Especially if there are any plans to enhance AwsCredentials class to support reading AWS credentials using WebIdentityTokenFileCredentialsProvider method.

I am having a use case that I have my service deployed on AWS EKS cluster which uses IRSA mechanism. Metadata service access is blocked. With the usage of AWSCredentialsProviderChain (especially WebIdentityTokenFileCredentialsProvider) I can get my AWS credentials using the AWS SDK, however the AWSCredentials class from the google-auth-library does not support this method of retrieving credentials. The AWSCredentials class (especially https://github.com/googleapis/google-auth-library-java/blob/577e9a52204b0d6026a302bb7efe2c6162d57945/oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java#L301) only scans the environment variables and if nothing is found it jumps to metadata service.

**Comment by edjeffreys** [comment]
Having the same issue as above from MSK - IMDS is unavailable, and ENV vars are not set. An easy solution would be to provide the `AwsSecurityCredentials` manually... https://github.com/googleapis/google-auth-library-java/pull/948

**Comment by sankalpkale88** [comment]
Hi @TimurSadykov , Any update on this ?

**Comment by TimurSadykov** [comment]
@lsirac could you please comment?

**Comment by lsirac** [comment]
We're aware of the gaps and will update when we add support for this.

**Comment by michalstefanext** [comment]
Hi @TimurSadykov @Isirac, do you have at least some rough ETA, when this could be introduced please?

**Comment by TimurSadykov** [comment]
@michalstefanext hopefully sometime next week

**Comment by lsirac** [comment]
> @michalstefanext hopefully sometime next week

Timur means next half :)

**Comment by aeitzman** [comment]
Added support for using a custom implementation to supply AWS security credentials in #1336, released in v1.23.0.
